README.md
A database specific redis pooling lib used by CartoDB node apps.

See http://cartodb.com

[![Build Status](https://travis-ci.org/CartoDB/node-redis-mpool.png)](http://travis-ci.org/CartoDB/node-redis-mpool)


NEWS
0.1.0 -- 2014-08-13

 - Switch to 3-clause BSD license (#3)
 - Upgraded dependencies
 - Implements validate method to take non connected clients from the pool.
 - Destroy calls to redis.end method to ensure socket is disconnected.
 - Client is returned on ready event after selecting the DB instead of using
   connect event.
 - New configuration allows to log acquire slow operations when it takes
   longer than a threshold to retrieve a client from the pool.

0.0.4 -- 2014-02-24

 - Add parameter to skip unwatch on release (#2)

0.0.3 -- 2013-12-11

 - Unwatch variables between connections (#1)

0.0.2 -- 2013-12-06

Fix dependencies (redis is not just for test...)

0.0.1 -- 2013-12-06

Initial release, spin off from node-cartodb-redis


test/support/config.js
var _ = require('underscore');

module.exports = function(opts) {

    var config = {
        redis_pool: {
            max: 10, 
            idleTimeoutMillis: 1, 
            reapIntervalMillis: 1,
            port: 6337
        }
    };

    _.extend(config,  opts || {});

    return config;
}();



test/redis_pool.test.js
var assert = require('assert')
  , Step = require('step')
  , _ = require('underscore')
  , RedisPool = require('../')
  , tests = module.exports = {};

suite('redis_pool', function() {

    // configure redis pool instance to use in tests
    var test_opts = require('./support/config').redis_pool;
    
    var redis_pool = new RedisPool(test_opts);

    test('RedisPool object exists', function(done){
      assert.ok(RedisPool);
      done();
    });
    
    test('RedisPool can create new redis_pool objects with default settings', function(done){
      new RedisPool();
      done();
    });
    
    test('RedisPool can create new redis_pool objects with specific settings', function(done){
      var redis_pool = new RedisPool(_.extend({host:'127.0.0.1', port: '6379'}, test_opts));
      done();
    });
    
    
    test('pool object has an acquire function', function(done){
      var found=false;
      var functions = _.functions(redis_pool);
      for (var i=0; i<functions.length; ++i) {
          if ( functions[i] == 'acquire' ) { found=true; break; }
      }
      assert.ok(found);
      done();
    });
    
    test('calling aquire returns a redis client object that can get/set', function(done){
      redis_pool.acquire(0, function(err, client){
        if ( err ) { done(err); return; }
        client.set("key","value");
        client.get("key", function(err,data){      
          assert.equal(data, "value");      
          redis_pool.release(0, client); // needed to exit tests
          done();
        })
      });    
    });
    
    test('calling aquire on another DB returns a redis client object that can get/set', function(done){
      redis_pool.acquire(2, function(err, client){
        if ( err ) { done(err); return; }
        client.set("key","value");
        client.get("key", function(err,data){      
          assert.equal(data, "value");      
          redis_pool.release(2, client); // needed to exit tests
          done();
        })
      });      
    });

    // See https://github.com/CartoDB/node-redis-mpool/issues/1
    test('calling release resets connection state', function(done){
      var client1, client2, tx1;
      Step(
        function getClient1() {
          redis_pool.acquire(0, this);
        },
        function getClient2(err, client) {
          if ( err ) throw err;
          client1 = client;
          redis_pool.acquire(0, this);
        },
        function regetClient1(err, client) {
          if ( err ) throw err;
          client2 = client;
          client1.WATCH('k');
          redis_pool.release(0, client1);
          client1 = null;
          redis_pool.acquire(0, this);
        },
        function startTransaction1(err, client) {
          if ( err ) throw err;
          client1 = client;
          // We expect this to be not watching now..
          tx1 = client1.MULTI();
          tx1.SET('x',1); // 'x' will be set to 1 only if we're not watching
          client2.SET('k',1, this);
        },
        function execTransaction1(err) {
          if ( err ) throw err;
          // This would fail if we're watching
          tx1.EXEC(this);
        },
        function checkTransaction(err, res) {
          if ( err ) throw err;
          assert.ok(res, "Transaction unexpectedly aborted"); // we expect to succeeded
          assert.equal(res.length, 1);
          return null;
        },
        function finish(err) {
          if ( client1 ) redis_pool.release(0, client1);
          if ( client2 ) redis_pool.release(0, client2);
          done(err);
        }
      );
    });

    test('log is called if elapsed time is above configured one', function(done) {
        var logWasCalled = false,
            elapsedThreshold = 25,
            enabledSlowPoolConfig = {
                slowPool: {
                    log: true,
                    elapsedThreshold: elapsedThreshold
                }
            };

        var times = 0;
        var dateNowFunc = Date.now;
        Date.now = function () {
            return times++ * elapsedThreshold * 2;
        };
        consoleLogFunc = console.log;
        console.log = function(what) {
            var whatObj;
            try {
                whatObj = JSON.parse(what);
            } catch (e) {
                // pass
            }
            logWasCalled = whatObj && whatObj.action && whatObj.action === 'adquire';
            consoleLogFunc.apply(console, arguments);
        }

        var redisPool = new RedisPool(_.extend(test_opts, enabledSlowPoolConfig));
        redisPool.acquire(0, function(err, client) {

            console.log = consoleLogFunc;
            Date.now = dateNowFunc;

            redisPool.release(0, client);
            assert.ok(logWasCalled);
            done();
        });
    });

});


index.js
var redis  = require('redis')
  , _      = require('underscore')
  , Pool   = require('generic-pool').Pool;

// constructor.
// 
// - `opts` {Object} optional config for redis and pooling
var RedisPool = function(opts) {
    var opts = opts || {};
    var defaults = {
        host: '127.0.0.1',
        port: '6379',
        max: 50,
        idleTimeoutMillis: 10000,
        reapIntervalMillis: 1000,
        unwatchOnRelease: true,
        name: 'default',
        log: false,
        slowPool: {
            log: false,
            elapsedThreshold: 25
        }
    };
    var options = _.defaults(opts, defaults)

    var me = {
        pools: {} // cached pools by DB name
    };

    var elapsedThreshold = options.slowPool.elapsedThreshold;
  
    // Acquire resource.
    //
    // - `database` {String} redis database name
    // - `callback` {Function} callback to call once acquired. Takes the form
    //   `callback(err, resource)`
    me.acquire = function(database, callback) {
        var pool = this.pools[database];
        if (!pool) {
        pool = this.pools[database] = this.makePool(database);
        }
        var startTime = Date.now();
        pool.acquire(function(err, client) {
            var elapsedTime = Date.now() - startTime;
            if (elapsedTime > elapsedThreshold) {
                log({db: database, action: 'adquire', elapsed: elapsedTime, waiting: pool.waitingClientsCount()});
            }
            callback(err, client);
        });
    };
  
    // Release resource.
    //
    // - `database` {String} redis database name
    // - `resource` {Object} resource object to release
    me.release = function(database, resource) {
        if ( options.unwatchOnRelease ) resource.UNWATCH();
        var pool = this.pools[database];
        if ( pool ) pool.release(resource);
    };
    
    // Factory for pool objects.
    me.makePool = function(database) {
        var pool = Pool({
            name: options.name + ':' + database,
            create: function(callback) {

                var callbackCalled = false;

                var client = redis.createClient(options.port, options.host);

                client.on('error', function (err) {
                    log({db: database, action: 'error', err: err.message});
                    if (!callbackCalled) {
                        callbackCalled = true;
                        callback(err, client);
                    }
                    client.end();
                });

                client.on('ready', function () {
                    client.select(database, function(err, res) {
                        if (!callbackCalled) {
                            callbackCalled = true;
                            callback(err, client);
                        }
                    });
                })
            },

            destroy: function(client) {
                client.quit();
                client.end();
            },

            validate: function(client) {
                return client && client.connected;
            },

            max: options.max,
            idleTimeoutMillis: options.idleTimeoutMillis,
            reapIntervalMillis: options.reapIntervalMillis,
            log: options.log
        });

        return pool;
    };

    function log(what) {
        if (options.slowPool.log) {
            console.log(JSON.stringify(_.extend({name: options.name}, what)));
        }
    }

    return me;
};

module.exports = RedisPool;
